reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java

```diff
@@ -0,0 +1,580 @@
+		t.setDaemon(true);
+		return t;
+	};
+
+	static final CachedService SHUTDOWN            = new CachedService(null);
+	static final int           DEFAULT_TTL_SECONDS = 60;
+
+	final ThreadFactory              factory;
+	final int                        ttlSeconds;
+	final int                        cap;
+	final Deque<CachedServiceExpiry> idleServicesWithExpiry;
+	final Queue<DeferredWorker>      deferredWorkers;
+	final Queue<CachedService>       allServices;
+	final ScheduledExecutorService   evictor;
+
+	volatile boolean                                        shutdown;
+	volatile int                                            remaining;
+	static final AtomicIntegerFieldUpdater<CappedScheduler> REMAINING =
+			AtomicIntegerFieldUpdater.newUpdater(CappedScheduler.class, "remaining");
+
+	CappedScheduler(int cap, ThreadFactory factory, int ttlSeconds) {
+		if (ttlSeconds < 0) {
+			throw new IllegalArgumentException("ttlSeconds must be positive, was: " + ttlSeconds);
+		}
+		this.ttlSeconds = ttlSeconds;
+		if (cap < 1) {
+			throw new IllegalArgumentException("cap must be strictly positive, was: " + cap);
+		}
+		this.cap = cap;
+		this.remaining = cap;
+		this.factory = factory;
+		this.idleServicesWithExpiry = new ConcurrentLinkedDeque<>();
+		this.deferredWorkers = new ConcurrentLinkedQueue<>();
+		this.allServices = new ConcurrentLinkedQueue<>();
+		this.evictor = Executors.newScheduledThreadPool(1, EVICTOR_FACTORY);
+		this.evictor.scheduleAtFixedRate(this::eviction,
+				ttlSeconds,
+				ttlSeconds,
+				TimeUnit.SECONDS);
+	}
+
+	/**
+	 * Instantiates the default {@link ScheduledExecutorService} for the CappedScheduler
+	 * ({@code Executors.newScheduledThreadPoolExecutor} with core and max pool size of 1).
+	 */
+	@Override
+	public ScheduledExecutorService get() {
+		ScheduledThreadPoolExecutor poolExecutor = new ScheduledThreadPoolExecutor(1, factory);
+		poolExecutor.setMaximumPoolSize(1);
+		poolExecutor.setRemoveOnCancelPolicy(true);
+		return poolExecutor;
+	}
+
+	@Override
+	public void start() {
+		throw new UnsupportedOperationException("Restarting not supported yet");
+	}
+
+	@Override
+	public boolean isDisposed() {
+		return shutdown;
+	}
+
+	@Override
+	public void dispose() {
+		if (shutdown) {
+			return;
+		}
+		shutdown = true;
+
+		evictor.shutdownNow();
+		idleServicesWithExpiry.clear();
+
+		CachedService cached;
+		while ((cached = allServices.poll()) != null) {
+			cached.exec.shutdownNow();
+		}
+	}
+
+	CachedService directPick() {
+		if (shutdown) {
+			return SHUTDOWN;
+		}
+		CachedService result;
+		//try to see if there is an idle worker
+		CachedServiceExpiry e = idleServicesWithExpiry.pollLast();
+		if (e != null) {
+			return e.cached;
+		}
+
+		if (REMAINING.decrementAndGet(this) < 0) {
+			//cap reached
+			REMAINING.incrementAndGet(this);
+			if (shutdown) {
+				return SHUTDOWN;
+			}
+			throw Exceptions.failWithRejected("cannot directly schedule task, cap of " + this.cap + " reached");
```

Comment:
don't we queue the tasks if we're at capacity?

Reply:
only for the cases where these tasks are submitted through a `Worker`

Reply:
Is there a strong reason for that? I am just thinking about some code that does `subscribeOn` inside a task and will fail, something I definitely wouldn't expect

Reply:
`subscribeOn` uses `Worker`

Reply:
the reason is practicality of implementation: all worker-based usage can be maintained consistent by providing the `DeferredWorker` that will queue the tasks. Free-form direct scheduling is much less obvious: should it queue in a common queue, a dedicated `DeferredWorker` that is only used for that purpose? round-robin to existing `DeferredWorker`s (with the caveat that this potentially impacts unrelated scheduling)? Reject? I simply chose the later because it was the simplest most deterministic solution.

Reply:
(just got that you were talking about `FluxSubscribeOnValue` @bsideup)

Reply:
I guess it could work that I `createWorker` on each direct schedule operation. Since the implementation calls `dispose()`, cancellation should correctly be propagated to the whole `Worker` through the returned `Disposable`.